File: backend/app/Models/Property.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasOne;

class Property extends Model
{
    protected $fillable = [
        'name',
        'description',
        'property_type_id',
        'user_id',
        'province_id',
        'city_id',
        'district_id',
        'subdistrict_id',
        'price',
        'address',
        'latitude',
        'longitude',
        'image',
        'capacity',
        'available_rooms',
        'rules',
        'isDeleted'
    ];

    protected $casts = [
        'price' => 'decimal:2',
        'latitude' => 'decimal:8',
        'longitude' => 'decimal:8',
        'isDeleted' => 'boolean',
        'meal_included' => 'boolean',
        'laundry_included' => 'boolean'
    ];

    // Relationships
    public function propertyType(): BelongsTo
    {
        return $this->belongsTo(PropertyType::class);
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function province(): BelongsTo
    {
        return $this->belongsTo(Province::class);
    }

    public function city(): BelongsTo
    {
        return $this->belongsTo(City::class);
    }

    public function district(): BelongsTo
    {
        return $this->belongsTo(District::class);
    }

    public function subdistrict(): BelongsTo
    {
        return $this->belongsTo(Subdistrict::class);
    }

    public function kostDetail(): HasOne
    {
        return $this->hasOne(KostDetail::class);
    }

    public function homestayDetail(): HasOne
    {
        return $this->hasOne(HomestayDetail::class);
    }

    public function rooms()
    {
        return $this->hasMany(Room::class);
    }

    // Accessors
    public function getImageUrlAttribute()
    {
        return $this->image ? asset('storage/'.$this->image) : asset('images/default-property.jpg');
    }

    // Scopes
    public function scopeActive($query)
    {
        return $query->where('isDeleted', false);
    }
}

File: frontend/lib/data/models/property.dart
import 'package:flutter/widgets.dart';

class Property {
  final BigInt id;
  final String name;
  final Text? description;
  final BigInt propertyTypeId;
  final BigInt userId;
  final BigInt provinceId;
  final BigInt cityId;
  final BigInt districtId;
  final BigInt subdistrictId;
  final double price;
  final Text? address;
  final double? latitude;
  final double? longitude;
  final String? image;
  final int? capacity;
  final int? availableRooms;
  final Text? rules;
  final bool isDeleted;

  // Optional: Nested relationships (lazy parsing)
  final String? imageUrl;
  final PropertyType? propertyType;
  final User? user;
  final Province? province;
  final City? city;
  final District? district;
  final Subdistrict? subdistrict;

  Property({
    required this.id,
    required this.name,
    this.description,
    required this.propertyTypeId,
    required this.userId,
    required this.provinceId,
    required this.cityId,
    required this.districtId,
    required this.subdistrictId,
    required this.price,
    this.address,
    this.latitude,
    this.longitude,
    this.image,
    this.capacity,
    this.availableRooms,
    this.rules,
    required this.isDeleted,
    this.imageUrl,
    this.propertyType,
    this.user,
    this.province,
    this.city,
    this.district,
    this.subdistrict,
  });

  factory Property.fromJson(Map<String, dynamic> json) {
    return Property(
      id: json['id'],
      name: json['name'],
      description: json['description'],
      propertyTypeId: json['property_type_id'],
      userId: json['user_id'],
      provinceId: json['province_id'],
      cityId: json['city_id'],
      districtId: json['district_id'],
      subdistrictId: json['subdistrict_id'],
      price: double.tryParse(json['price'].toString()) ?? 0.0,
      address: json['address'],
      latitude: json['latitude'] != null
          ? double.tryParse(json['latitude'].toString())
          : null,
      longitude: json['longitude'] != null
          ? double.tryParse(json['longitude'].toString())
          : null,
      image: json['image'],
      capacity: json['capacity'],
      availableRooms: json['available_rooms'],
      rules: json['rules'],
      isDeleted: json['isDeleted'] ?? false,
      imageUrl: json['image_url'], // Dari accessor Laravel
      // Optional parsing for relationships
      propertyType: json['property_type'] != null
          ? PropertyType.fromJson(json['property_type'])
          : null,
      user: json['user'] != null ? User.fromJson(json['user']) : null,
      province: json['province'] != null
          ? Province.fromJson(json['province'])
          : null,
      city: json['city'] != null ? City.fromJson(json['city']) : null,
      district: json['district'] != null
          ? District.fromJson(json['district'])
          : null,
      subdistrict: json['subdistrict'] != null
          ? Subdistrict.fromJson(json['subdistrict'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'description': description,
      'property_type_id': propertyTypeId,
      'user_id': userId,
      'province_id': provinceId,
      'city_id': cityId,
      'district_id': districtId,
      'subdistrict_id': subdistrictId,
      'price': price,
      'address': address,
      'latitude': latitude,
      'longitude': longitude,
      'image': image,
      'capacity': capacity,
      'available_rooms': availableRooms,
      'rules': rules,
      'isDeleted': isDeleted,
    };
  }
}
